### ISSUE TITLE
"Unknown" Zoom versions

### ISSUE THREAD

**Issue by iainbrighton** [aaronparker/Evergreen#35]
**Describe the bug**
When running `Get-Zoom` the versions for most downloads is listed as `Unknown`
```
PS C:\Users\Iain\Dropbox\Apps\PowerShell\6.1.1> Get-Zoom

Version        Platform Type    URI
-------        -------- ----    ---
Unknown        Windows  Outlook https://zoom.us/client/latest/ZoomOutlookPluginSetup.msi
4.6.17409.0120 Windows  EXE     https://d11yldzmag5yn.cloudfront.net/prod/4.6.17409.0120/ZoomInstaller.exe
Unknown        Windows  Lync    https://zoom.us/client/latest/ZoomLyncPluginSetup.msi
Unknown        Windows  Rooms   https://zoom.us/client/latest/ZoomRooms.exe
4.7.52180.0404 Windows  Notes   https://d11yldzmag5yn.cloudfront.net/prod/4.7.52180.0404/ZoomNotesPlugi...
Unknown        Windows  MSI     https://zoom.us/client/latest/ZoomInstallerFull.msi
Unknown        Citrix   HPTh... https://d11yldzmag5yn.cloudfront.net/prod/vdi/zoomcitrixplugin-hp_amd64...
Unknown        Citrix   eLux... https://d11yldzmag5yn.cloudfront.net/prod/vdi/zoomcitrixplugin_library_...
Unknown        Citrix   eLux... https://d11yldzmag5yn.cloudfront.net/prod/vdi/zoomcitrixplugin_library_...
Unknown        Citrix   HPTh... https://d11yldzmag5yn.cloudfront.net/prod/vdi/zoomcitrixplugin-hp_i386.deb
Unknown        Citrix   Wind... https://d11yldzmag5yn.cloudfront.net/prod/vdi/ZoomCitrixHDXMediaPlugin.msi
Unknown        Citrix   IGEL... https://d11yldzmag5yn.cloudfront.net/prod/vdi/zoomcitrixplugin-igel_64.zip
Unknown        Citrix   Wind... https://d11yldzmag5yn.cloudfront.net/prod/vdi/ZoomInstallerVDI.msi
Unknown        VMware   Client  https://d11yldzmag5yn.cloudfront.net/prod/vdi/ZoomVmwareMediaPlugin.msi
```

I have confirmed this is the current behaviour in Windows PowerShell 5.1 and PowerShell Core 6.1.1 on Windows 10 1909. It appears that the [Resolve-RedirectedUri](https://github.com/aaronparker/Evergreen/blob/master/Evergreen/Private/Resolve-RedirectedUri.ps1) is having problems following some of the links.

I'm not sure whether it helps or not, but this code (I am only interested in the MSI versions!) functions on both Windows PowerShell 5.1 and PowerShell Core 6.1.1 on Windows 10 1909:
```
function Resolve-Uri
{
    [CmdletBinding()]
    param (
        [Parameter(Mandatory, ValueFromPipeline, ValueFromPipelineByPropertyName)]
        [System.String] $Uri
    )
    process
    {
        try
        {
            $httpWebRequest = [System.Net.WebRequest]::Create($uri)
            $httpWebRequest.MaximumAutomaticRedirections = 3
            $httpWebRequest.AllowAutoRedirect = $true
            $webResponse = $httpWebRequest.GetResponse()
            Write-Output -InputObject $webResponse.ResponseUri.AbsoluteUri
        }
        catch
        {
            throw $_
        }
        finally
        {
            $webResponse.Dispose()
        }
    }
}

function Get-ZoomEx
{
    [CmdletBinding()]
    param ( )
    process
    {
        $downloads = @{
            Client = 'http://zoom.us/client/latest/ZoomInstallerFull.msi'
            Outlook = 'http://zoom.us/client/latest/ZoomOutlookPluginSetup.msi'
            Notes = 'https://zoom.us/client/latest/ZoomNotesPluginSetup.msi'
            Lync = 'http://zoom.us/client/latest/ZoomLyncPluginSetup.msi'
            Rooms = 'http://zoom.us/client/latest/ZoomRoomsInstaller.msi'
        }

        foreach ($download in $downloads.GetEnumerator())
        {
            $version = 'Unknown'
            $uri = Resolve-Uri -Uri $download.Value
            $regexMatch = [Regex]::Match($uri, '(?<=/)\d+(\.\d+){2,}(?=\.\d+/)')
            if ($regexMatch.Success -eq $true)
            {
                $version = $regexMatch.Value
            }
            [PSCustomObject] @{
                Version = $version
                Product = $download.Key
                Uri = Resolve-Uri -Uri $download.Value
            }
        }
    }
}

PS C:\Users\Iain\Dropbox\Apps\PowerShell\6.1.1> Get-ZoomEx -Verbose

Version   Product Uri
-------   ------- ---
4.6.17409 Client  https://d11yldzmag5yn.cloudfront.net/prod/4.6.17409.0120/ZoomInstallerFull.msi
4.6.1244  Rooms   https://d11yldzmag5yn.cloudfront.net/prod/4.6.1244.0112/ZoomRoomsInstaller.msi
4.7.52180 Notes   https://d11yldzmag5yn.cloudfront.net/prod/4.7.52180.0404/ZoomNotesPluginSet...
4.8.17303 Outlook https://d11yldzmag5yn.cloudfront.net/prod/4.8.17303.0117/ZoomOutlookPluginS...
4.8.13207 Lync    https://d11yldzmag5yn.cloudfront.net/prod/4.8.13207.1127/ZoomLyncPluginSetu...
```
Feel free to use whatever helps!

Thanks,
Iain

**Comment by aaronparker** [comment]
Thanks. I'll take a look. I'll have to update `Resolve-RedirectedUri` perhaps and re-test across all functions first.

**Comment by aaronparker** [comment]
@iainbrighton is `(?<=/)\d+(\.\d+){2,}(?=\.\d+/)` a better way to match a version number than what I've been using previously?

**Comment by aaronparker** [comment]
fixed in 2002.120

**Comment by iainbrighton** [comment]
Not sure! It just uses positive look-behind (?<=) and look-ahead (?=) matches.

The (?<=/) looks for a forward slash. If it doesn’t find one, the main capture group is not enumerated. The benefit is that this is not included in the resulting match(es).

The (?=\.\d+/) look for a period, a number and a forward slash. The benefit is that this is also not included in the resulting match(es), effectively stripping the last version.

Here are some examples – notice that the 3-part version number does NOT match.

C:\..\Dropbox\PowerShell> '/1.2.3/' -match '(?<=/)\d+(\.\d+){2,}(?=\.\d+/)'; $matches[0]
False

C:\..\Dropbox\PowerShell> '/1.2.3.4/' -match '(?<=/)\d+(\.\d+){2,}(?=\.\d+/)'; $matches[0]
True
1.2.3

C:\..\Dropbox\PowerShell> '/1.2.3.4.5/' -match '(?<=/)\d+(\.\d+){2,}(?=\.\d+/)'; $matches[0]
True
1.2.3.4

C:\..\Dropbox\PowerShell> '/1.2.3.4.5.6/' -match '(?<=/)\d+(\.\d+){2,}(?=\.\d+/)'; $matches[0]
True
1.2.3.4.5

I don’t know whether you wanted to return all version number parts. In my implementation, I didn’t! If you want all the version number bits you could use (?<=/)\d+(\.\d+){2,}(?=/)

C:\..\Dropbox\PowerShell> '/1.2.3/' -match '(?<=/)\d+(\.\d+){2,}(?=/)'; $matches[0]
True
1.2.3

C:\..\Dropbox\PowerShell> '/1.2.3.4/' -match '(?<=/)\d+(\.\d+){2,}(?=/)'; $matches[0]
True
1.2.3.4

C:\..\Dropbox\PowerShell> '/1.2.3.4.5/' -match '(?<=/)\d+(\.\d+){2,}(?=/)'; $matches[0]
True
1.2.3.4.5

C:\..\Dropbox\PowerShell> '/1.2.3.4.5.6/' -match '(?<=/)\d+(\.\d+){2,}(?=/)'; $matches[0]
True
1.2.3.4.5.6

For completeness, a simple 3 part (or more) version match regex could simply look like this: \d+(\.\d+){2,}

Thanks,
Iain
